RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitMultiFitAlgo_gpu.cu

```diff
@@ -0,0 +1,1033 @@
+                                    break;
+                                }
+                            } 
+                        }
+                        float correctedTime = (crh.timeMax-5) * clockToNsConstant + itimeconst + offsetTime;
+                        float cterm         = EBtimeConstantTerm;
+                        float sigmaped      = pedRMSVec[0];  // approx for lower gains
+                        float nterm         = EBtimeNconst*sigmaped/uncalibRecHit.amplitude();
+                        float sigmat        = std::sqrt( nterm*nterm  + cterm*cterm   );
+                        if ( ( correctedTime > sigmat*outOfTimeThreshP )   ||
+                             ( correctedTime < -sigmat*outOfTimeThreshM ))
+                        {
+                            uncalibRecHit.setFlagBit( EcalUncalibratedRecHit::kOutOfTime );
+                        }
+                    }
+                } // if not barrel
+            } // if timeAlgo_ ==
+        }
+    }
+}
+
+void scatter(host_data& h_data, device_data& d_data) {
+
+/*
+void scatter(EcalDigiCollection const& digis,
+             EcalUncalibratedRecHitCollection& rechits,
+             std::vector<EcalPedestal> const& vpedestals,
+             std::vector<EcalMGPAGainRatio> const& vgains,
+             std::vector<EcalXtalGroupId> const& vxtals,
+             std::vector<EcalPulseShape> const& vpulses,
+             std::vector<EcalPulseCovariance> const& vcovariances,
+             SampleMatrixGainArray const& noisecors,
+             device_data &d_data) {
+*/
+    auto const& ids = h_data.digis->ids();
+    auto const& digis_data = h_data.digis->data();
+    using digis_type = std::vector<uint16_t>;
+    using dids_type = std::vector<uint32_t>;
+    
+    //
+    // TODO: remove per event alloc/dealloc -> do once at the start
+    //
+    /*
+    cudaMalloc((void**)&d_digis_data,
+        digis_data.size() * sizeof(digis_type::value_type));
+    cudaMalloc((void**)&d_ids,
+        ids.size() * sizeof(dids_type::value_type));
+    cudaMalloc((void**)&d_pedestals,
+        vpedestals.size() * sizeof(EcalPedestal));
+    cudaMalloc((void**)&d_gains, 
+        vgains.size() * sizeof(EcalMGPAGainRatio));
+    cudaMalloc((void**)&d_xtals,
+        vxtals.size() * sizeof(EcalXtalGroupId));
+    cudaMalloc((void**)&d_shapes,
+        vpulses.size() * sizeof(EcalPulseShape));
+    cudaMalloc((void**)&d_covariances,
+        vcovariances.size() * sizeof(EcalPulseCovariance));
+    cudaMalloc((void**)&d_rechits,
+        rechits.size() * sizeof(EcalUncalibratedRecHit));
+    cudaMalloc((void**)&d_noisecors,
+        noisecors.size() * sizeof(SampleMatrix));
+    ecal::cuda::assert_if_error();
+    */
+
+    // 
+    // copy to the device
+    // TODO: can conditions be copied only once when updated?
+    //
+    cudaMemcpy(d_data.digis_data, digis_data.data(),
+        digis_data.size() * sizeof(digis_type::value_type),
+        cudaMemcpyHostToDevice);
```

Comment:
Eventually all these need to be async. It follows that in all cases the host memory must be pinned (i.e. allocated with `cudaMallocHost()`/`cudaHostAlloc()`).